jmclient/jmclient/wallet.py

```diff
@@ -1032,6 +999,501 @@ def close(self):
+        about a PSBT_Output object.
+        """
+        assert isinstance(psbt_output, btc.PSBT_Output)
+        outdict = {}
+        if psbt_output.index is not None:
+            outdict["output-index"] = psbt_output.index
+
+        if psbt_output.derivation_map:
+            # See note to derivation map in hr_psbt_in()
+            outdict["derivation-map"] = {bintohex(k): bintohex(v.serialize(
+                )) for k, v in psbt_output.derivation_map.items()}
+
+        if psbt_output.redeem_script:
+            outdict["redeem-script"] = bintohex(psbt_output.redeem_script)
+        if psbt_output.witness_script:
+            outdict["witness-script"] = bintohex(psbt_output.witness_script)
+
+        if psbt_output.proprietary_fields:
+            outdict["proprietary-fields"] = str(psbt_output.proprietary_fields)
+        if psbt_output.unknown_fields:
+            outdict["unknown-fields"] = str(psbt_output.unknown_fields)
+
+        return outdict
+
+    @staticmethod
+    def witness_utxos_to_psbt_utxos(utxos):
+        """ Given a dict of utxos as returned from select_utxos,
+        convert them to the format required to populate PSBT inputs,
+        namely CTxOut. Note that the non-segwit case is different, there
+        you should provide an entire CMutableTransaction object instead.
+        """
+        return [btc.CMutableTxOut(v["value"],
+                                  v["script"]) for _, v in utxos.items()]
+
+    def create_psbt_from_tx(self, tx, spent_outs=None):
+        """ Given a CMutableTransaction object, which should not currently
+        contain signatures, we create and return a new PSBT object of type
+        btc.PartiallySignedTransaction.
+        Optionally the information about the spent outputs that is stored
+        in PSBT_IN_NONWITNESS_UTXO, PSBT_IN_WITNESS_UTXO and PSBT_IN_REDEEM_SCRIPT
+        can also be provided, one item per input, in the tuple (spent_outs).
+        These objects should be either CMutableTransaction, CTxOut or None,
+        Note that redeem script information cannot be provided for inputs which
+        we don't own.
+        """
+        # TODO: verify tx contains no signatures as a sanity check?
+        new_psbt = btc.PartiallySignedTransaction(unsigned_tx=tx)
+        if spent_outs is None:
+            # user has not provided input script information; psbt
+            # will not yet be usable for signing.
+            return new_psbt
+        for i, txinput in enumerate(new_psbt.inputs):
+            if spent_outs[i] is None:
+                # as above, will not be signable in this case
+                continue
+            if isinstance(spent_outs[i], (btc.CTransaction, btc.CTxOut)):
+                # note that we trust the caller to choose Tx vs TxOut as according
+                # to non-witness/witness. Note also that for now this mixin does
+                # not attempt to provide previous tx (second argument) for witness
+                # case.
```

Comment:
Second argument is not previous tx. It is `unsigned_tx` of psbt. You have it right here in `new_psbt.unsigned_tx`. This argument is required because I did not want to link back `PSBT_Input` to the parent psbt, so `PSBT_Input` only requires `unsigned_tx` on creation and uses it for checking and getting `prevout.n` and `prevout.hash` from appropriate input. It uses it for the same purpose in `set_utxo()`.

Reply:
Oops, will fix comment in next commit, thanks.

Reply:
Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/536/commits/e9367fd755bfbb94c5bb1b73a07a2550536b942c